File: src/main/java/com/codingspezis/android/metalonly/player/plan/PlanEntryToItemConverter.java
package com.codingspezis.android.metalonly.player.plan;

import com.codingspezis.android.metalonly.player.fragments.PlanFragment;

import org.androidannotations.annotations.EBean;
import org.androidannotations.annotations.res.StringArrayRes;
import org.androidannotations.annotations.res.StringRes;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.GregorianCalendar;

@EBean
public class PlanEntryToItemConverter {

    @StringRes
    String plan;
    @StringArrayRes
    String[] days;


    public ArrayList<PlanItem> convertToPlan(ArrayList<PlanData> listEvents) {
        // TODO refactor this method
        ArrayList<PlanItem> listItems = new ArrayList<PlanItem>();
        Calendar cal = new GregorianCalendar();

        int day = 0;

        PlanSectionItem nextDaySection = new PlanSectionItem(days[day]);

        listItems.add(nextDaySection);

        for (int i = 0; i < listEvents.size(); i++) {
            PlanData d = listEvents.get(i);
            listItems.add(new PlanRealEntryItem(d));
            if (hasNextListItem(listEvents, i)) {
                PlanData nextItem = listEvents.get(i + 1);
                if (notOnSameDay(d, nextItem)) {
                    day++;
                    nextDaySection = new PlanSectionItem(days[day]);
                    listItems.add(nextDaySection);
                    if (isToday(day)){
                        // TODO setTodayListStartIndex(listItems.size());
                    }
                    int dayOfWeek = (cal.get(Calendar.DAY_OF_WEEK) + 5) % 7;
                    if (day == dayOfWeek) {
                        final int pos = listItems.size() - 1;
                        // TODO planFragment.getList().setSelection(pos);
                    }
                }
            }
        }
        return listItems;
    }


    private boolean isToday(int dayIndex) {
        Calendar cal = new GregorianCalendar();
        return (cal.get(Calendar.DAY_OF_WEEK) + 5) % 7 == dayIndex;
    }

    private boolean notOnSameDay(PlanData d, PlanData nextItem) {
        return !d.sameDay(nextItem);
    }

    private boolean hasNextListItem(ArrayList<PlanData> listEvents, int i) {
        return i < listEvents.size() - 1;
    }
}

File: src/main/kotlin/com/codingspezis/android/metalonly/player/plan/PlanEntryToItemConverter.kt
package com.codingspezis.android.metalonly.player.plan

import org.androidannotations.annotations.EBean
import org.androidannotations.annotations.res.StringArrayRes
import org.androidannotations.annotations.res.StringRes

import java.util.ArrayList
import java.util.Calendar
import java.util.GregorianCalendar

@EBean
open class PlanEntryToItemConverter {

    @JvmField
    @StringRes
    protected var plan: String? = null

    @JvmField
    @StringArrayRes
    protected var days: Array<String>? = null

    internal var todayStartIndex: Int = 0

    fun convertToPlan(listEvents: ArrayList<ShowInformation>): ArrayList<PlanItem> {
        // TODO refactor this method
        val listItems = ArrayList<PlanItem>()

        var dayIndex = 0

        listItems.add(PlanSectionItem(days!![dayIndex]))

        for (i in listEvents.indices) {
            val currentShow = listEvents[i]
            listItems.add(PlanRealEntryItem(currentShow))

            if (hasNextListItem(listEvents, i)) {
                val nextShow = listEvents[i + 1]
                val notOnSameDay = getDayIndex(currentShow) != getDayIndex(nextShow)
                if (notOnSameDay) {
                    dayIndex++
                    listItems.add(PlanSectionItem(days!![dayIndex]))
                    if (isToday(dayIndex)) {
                        todayStartIndex = listItems.size - 1
                    }
                }
            }
        }
        return listItems
    }

    private fun isToday(dayIndex: Int): Boolean {
        val cal = GregorianCalendar()
        return (cal.get(Calendar.DAY_OF_WEEK) + 5) % 7 == dayIndex
    }

    private fun getDayIndex(d: ShowInformation): Int {
        val cal = Calendar.getInstance()
        cal.time = d.startDate
        return cal.get(Calendar.DAY_OF_WEEK)
    }

    private fun hasNextListItem(listEvents: ArrayList<ShowInformation>, i: Int): Boolean {
        return i < listEvents.size - 1
    }

    fun todayStartIndex(): Int {
        return todayStartIndex
    }
}